File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/CqlFunctionOverloadTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.math.BigDecimal;
import org.junit.jupiter.api.Test;

class CqlFunctionOverloadTest extends CqlTestBase {

    @Test
    void function_overloads() {
        var results = engine.evaluate(toElmIdentifier("FunctionOverloadTest"));
        var value = results.forExpression("TestAnyFunctionWithInteger").value();
        assertThat(value, is(1));

        value = results.forExpression("TestAnyFunctionWithString").value();
        assertThat(value, is("joe"));

        value = results.forExpression("TestAnyFunctionWithDecimal").value();
        assertThat(value, is(new BigDecimal("12.3")));

        value = results.forExpression("TestAnyFunctionWithNoArgs").value();
        assertThat(value, is("any"));
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/execution/CqlFunctionOverloadTest.kt
package org.opencds.cqf.cql.engine.execution

import java.math.BigDecimal
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test

internal class CqlFunctionOverloadTest : CqlTestBase() {
    @Test
    fun function_overloads() {
        val results = engine.evaluate(toElmIdentifier("FunctionOverloadTest"))
        var value = results.forExpression("TestAnyFunctionWithInteger")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(1))

        value = results.forExpression("TestAnyFunctionWithString")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`("joe"))

        value = results.forExpression("TestAnyFunctionWithDecimal")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(BigDecimal("12.3")))

        value = results.forExpression("TestAnyFunctionWithNoArgs")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`("any"))
    }
}
